### PR TITLE
Rspec testing

### PR DESCRIPTION
**Pull request by mcrd25**
Testing for our custom enumerable functions using RSpec

### CODE REVIEW

File: spec/my_enumerables_spec.rb [hunk 1]
```diff
@@ -0,0 +1,226 @@
+require './my_enumerables.rb'
+
+describe Enumerable do
+  context '#my_each' do
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: You use the same array `[1,2,3]` in many of the test cases. [Use let instead](http://www.betterspecs.org/#let)

File: spec/my_enumerables_spec.rb [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      expect { [1, 2, 3].my_each { |i| print i * 2 } }.to output(/246/).to_stdout
+    end
+
+    it 'returns an Enumerator when no block given' do
+      expect([1, 2, 3].my_each.is_a?(Enumerator)).to be true
+    end
+
+    it 'all above-mentioned tests match the official enumerable method counterpart' do
+      expect([1, 2, 3].my_each { |i| i * 2 }).to eq([1, 2, 3].each { |i| i * 2 })
+      expect { [1, 2, 3].each { |i, index| print i * 2 } }.to output(/246/).to_stdout
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: Here you are only testing the original `each` method. This test case is not necessary

File: spec/my_enumerables_spec.rb (hunk 1 above)
Comment (heyapricot) [reply spec/my_enumerables_spec.rb]: Same with repeated blocks, you can store it with `let` and use it when necessary

File: spec/my_enumerables_spec.rb [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+      
+    end
+
+    it 'returns an Enumerator when no block given' do
+      expect([1, 2, 3].my_each_with_index.is_a?(Enumerator)).to be true
+    end
+
+    it 'all above-mentioned tests match the official enumerable method counterpart' do
+      expect([1, 2, 3].my_each_with_index { |i| i * 2 }).to eq([1, 2, 3].each_with_index { |i| i * 2 })
+      expect { [1, 2, 3].each_with_index {|i, index| print i, index} }.to output(/102132/).to_stdout
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: Also not necessary

File: spec/my_enumerables_spec.rb (hunk 2 above)
Comment (heyapricot) [reply spec/my_enumerables_spec.rb]: I'm referring to line 19 only

File: spec/my_enumerables_spec.rb (hunk 3 above)
Comment (heyapricot) [reply spec/my_enumerables_spec.rb]: Referring to line 40 only

File: spec/my_enumerables_spec.rb [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      expect(%w[ant bear cat].my_none?(/d/)).to eq(%w[ant bear cat].none?(/d/))
+    end
+  end
+  
+  context '#my_count' do
+    it 'returns array length when no block given' do
+      expect([1, 2, 3, 4].my_count).to eq 4
+    end
+
+    it 'returns number of items in array when item is an argument' do
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: it returns the number of items that match the argument

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Very nice looking tests 🤩 🧪
You did a great job with these tests! I can tell you put a lot of effort. Hope you are falling in love with TDD 🧪❤️

My only comment for improvement is that you can use `let` to define a test variable and share it among your test cases. (You can also store a block). 

Add the use of let to your test cases and we'll be done :)

File: spec/my_enumerables_spec.rb (hunk 4 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Please refer to Commit No. #6

File: spec/my_enumerables_spec.rb (hunk 3 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Please refer to Commit No. #6

File: spec/my_enumerables_spec.rb (hunk 2 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Please refer to Commit No. #6

File: spec/my_enumerables_spec.rb (hunk 1 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Please refer to Commit No. #6. Alvaro, could you please advise on storing a block within a let. Done some research on that and have tried to no avail.

Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: A proc within a let where possible with latest commit see below

File: spec/my_enumerables_spec.rb [hunk 5]
```diff
@@ -1,69 +1,72 @@
...
     it 'all above-mentioned tests match the official enumerable method counterpart' do
-      expect([1, 2, 3, 4].my_select(&:even?)).to eq([1, 2, 3, 4].select(&:even?))
-      expect([1, 2, 3].my_select.is_a?(Enumerator)).to eq([1, 2, 3].select.is_a?(Enumerator))
+      expect(test_arr.my_select(&:even?)).to eq(test_arr.select(&:even?))
+      expect(test_arr.my_select.is_a?(Enumerator)).to eq(test_arr.select.is_a?(Enumerator))
     end
   end
 
-  context '#my_all?' do 
+  let(:ant_bear_cat) { %w[ant bear cat] }
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: If this element is going to be available for all the describe blocks inside `describe Enumerable` why not define it in L6 after `test_block`?

Either do that or put the `let(:ant_bear_cat)` inside the `describe '#my_all?'`

File: spec/my_enumerables_spec.rb [hunk 6]
```diff
@@ -83,26 +86,29 @@
...
       expect([].my_all?).to eq([].all?)
       expect([1, 2i, 3.14].my_all?(Numeric)).to eq([1, 2i, 3.14].my_all?(Numeric))
-      expect(%w[ant bear cat].my_all?(/t/)).to eq(%w[ant bear cat].all?(/t/))
+      expect(ant_bear_cat.my_all?(/t/)).to eq(%w[ant bear cat].all?(/t/))
     end
   end
 
-  context '#my_any' do 
+  let(:len_block_three) { |word| word.length >= 3 }
+  let(:len_block_five) { |word| word.length >= 5 }
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: Forgot to store the block as a `proc`

File: spec/my_enumerables_spec.rb [hunk 7]
```diff
@@ -83,26 +86,29 @@
...
+
+  describe '#my_any' do 
     it 'returns true when at least one item matches given condition' do
-      expect(%w[ant bear cat].my_any? { |word| word.length >= 3 }).to be true
+      expect(ant_bear_cat.my_any? { |word| word.length >= 3 }).to be true
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: After changing to `let(:len_block_three) { proc |word| word.length >= 3 }` tested with `expect(ant_bear_cat.my_any?(&len_block_three)).to be true` and it worked fine for me

File: spec/my_enumerables_spec.rb [hunk 8]
```diff
@@ -143,84 +149,84 @@
...
     end
   end
 
-  context '#my_map' do
+  let(:test_range) { (1..4) }
+
+  describe '#my_map' do
     it 'returns a modified array resulting from block operation' do
-      expect([1, 2, 3, 4].my_map { |i| i * i }).to eq [1, 4, 9, 16]
+      expect(test_arr.my_map { |i| i * i }).to eq([1, 4, 9])
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: Why is `[1,4,9]` still hardcoded?

File: spec/my_enumerables_spec.rb [hunk 9]
```diff
@@ -118,20 +124,20 @@
     end
 
     it 'all above-mentioned tests match the official enumerable method counterpart' do
-      expect(%w[ant bear cat].my_any? { |word| word.length >= 3 }).to eq(%w[ant bear cat].any? { |word| word.length >= 3 })
-      expect(%w[ant bear cat].my_any? { |word| word.length >= 5 }).to eq(%w[ant bear cat].any? { |word| word.length >= 5 })
+      expect(ant_bear_cat.my_any? { |word| word.length >= 3 }).to eq(ant_bear_cat.any? { |word| word.length >= 3 })
+      expect(ant_bear_cat.my_any? { |word| word.length >= 5  }).to eq(ant_bear_cat.any? { |word| word.length >= 5 })
       expect([nil, true, false].my_any?).to eq([nil, true, false].any?)
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb]: Why is `[nil, true, false]` still hardcoded?

File: spec/my_enumerables_spec.rb (hunk 8 above)
Comment (heyapricot) [reply spec/my_enumerables_spec.rb]: I really like the idea of storing the blocks as procs for these tests 🙈

File: spec/my_enumerables_spec.rb [hunk 10]
```diff
@@ -143,84 +149,84 @@
...
     end
 
     it 'returns modified array when given a range with a block' do
-      expect((1..4).my_map { |i| i * i }).to eq [1, 4, 9, 16]
+      expect(test_range.my_map { |i| i * i }).to eq([1, 4, 9, 16])
     end
 
     it 'returns array of strings when given array of integers' do
-      expect([1, 2].my_map(&:to_s)).to eq ['1', '2']
+      expect(test_arr.my_map(&:to_s)).to eq(['1', '2', '3'])
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb:195]: No hardcoded elements please

Comment (heyapricot) [reply spec/my_enumerables_spec.rb:195]: No `[1,2,3]` directly

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Avoid hardcoded elements as much as you can
The idea behind using let is that we can have a shared variable for test cases and do some memory optimizations. It also allows us to make our tests more intense just by changing a variable
E.g. Changing the array from `[1,2,3]` to a larger range

Speaking of larger range:
Something I like to do is have constants at the beggining of the spec that I can change easily to test with more elements. E.g.
```
ARRAY_LENGTH = 10 
let(:test_arr) { (1..ARRAY_LENGTH).to_a }
```

File: spec/my_enumerables_spec.rb (hunk 5 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: I was following one of Uncle Bob's clean code principles of having variables closer to where they are used. One tends to forget about it and would have to scroll, that is the reason why I placed it there. I've placed them on top though, and I can also see the benefits of doing that.

File: spec/my_enumerables_spec.rb (hunk 6 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Yes, this was left overs from yesterday's tryouts. I've corrected it and implemented your suggestion.

File: spec/my_enumerables_spec.rb (hunk 7 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Thank you. I tried similar approaches yesterday to no avail. Your suggestion has been implemented.

File: spec/my_enumerables_spec.rb (hunk 8 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Done

File: spec/my_enumerables_spec.rb (hunk 9 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb]: Changed

File: spec/my_enumerables_spec.rb (hunk 10 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb:195]: Done

File: spec/my_enumerables_spec.rb [hunk 11]
```diff
@@ -181,15 +184,13 @@
     end
   end
 
-  let(:test_range) { (1..4) }
-
   describe '#my_map' do
     it 'returns a modified array resulting from block operation' do
-      expect(test_arr.my_map { |i| i * i }).to eq([1, 4, 9])
+      expect(test_arr.my_map(&test_block)).to eq([2, 4, 6])
```
Comment (heyapricot) [inline spec/my_enumerables_spec.rb:187]: Not crazy about the `eq` having hardcoded values. You'll get better at this with practice

**Review by heyapricot — APPROVED** [review]
# *Let*'s approve it

You did an excellent job with the tests and learned about let in the process. Isn't that nice?
No need from my side for further review. Moving on to the next project 🚀

- P.S. Consider adding .vscode to your .gitignore file

File: spec/my_enumerables_spec.rb (hunk 11 above)
Comment (adriaanbd) [reply spec/my_enumerables_spec.rb:187]: That's the plan :)

**Comment by mcrd25** [comment]
Thanks! Will 🙂

**Pull request merged**